### PR TITLE
Shadow replicas on shared filesystems

### PR DESCRIPTION
**Pull request by dakrone**
These commits add the shadow replicas feature for use on shared filesystems
(it does not include segment replication for non-shared filesystems yet).

If we assume that the data in the index path will already be shared across
multiple nodes, we can create and index with shadow replicas, where each replica
shard simply contains an `IndexReader` that periodically refreshes to pick up
new segments.

All indexing operations will be executed on the primary shard, and will not be
replicated to each replica, since the data will be replicated in a different
way.

During this phase, creating an index with `index.shadow_replicas: true` and
`number_of_replicas` greater than 0 will cause operations not to undergo
replication to replica shards. An index can have either regular replicas or
shadow replicas; they are mutually exclusive for an index. The
`index.shadow_replicas` setting is set at index creation time and cannot be
changed dynamically.

The Elasticsearch cluster will still detect the loss of a primary shard, and
transform the replica into a primary in this situation. This transformation will
take slightly longer, since no `IndexWriter` will be maintained for each shadow
replica.

In order to ensure the data is being synchronized in a fast enough manner, The
user will need to tune the flush threshold for the index to a desired number. A
flush is needed to fsync segment files to disk, so they will be visible to all
other replica nodes. Users should test what flush threshold levels they are
comfortable with, as increased flushing can impact indexing performance. This
testing can be performed at any time, there is no need to wait for this feature
to be available first.

Once segments are available on the filesystem where the shadow replica resides,
a regular refresh (governed by the `index.refresh_interval`) can be used to make
the new data searchable.

See #8976 for the overall shadow replica plan

### CODE REVIEW

File: docs/reference/indices/shadow-replicas.asciidoc [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+--------------------------------------------------
+node.enable_custom_paths: true
+--------------------------------------------------
+
+You can then create an index with a custom data path, where each node will use
+this path for the data:
+
+[source,js]
+--------------------------------------------------
+curl -XPUT 'localhost:9200/my_index/_settings' -d '
```
Comment (imotov) [inline docs/reference/indices/shadow-replicas.asciidoc]: That should be `curl -XPUT 'localhost:9200/my_index`.

File: docs/reference/indices/shadow-replicas.asciidoc [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+{
+    "index" : {
+        "number_of_shards" : 1
```
Comment (imotov) [inline docs/reference/indices/shadow-replicas.asciidoc]: Missing comma at the end.

File: src/main/java/org/elasticsearch/indices/recovery/RecoveriesCollection.java [hunk 3]
```diff
@@ -20,6 +20,7 @@
 package org.elasticsearch.indices.recovery;
 
 import org.elasticsearch.ElasticsearchTimeoutException;
+import org.elasticsearch.cluster.metadata.IndexMetaData;
```
Comment (imotov) [inline src/main/java/org/elasticsearch/indices/recovery/RecoveriesCollection.java]: Unused

**Comment by imotov** [comment]
I left a couple of cosmetic comments. LGTM.

File: docs/reference/indices/shadow-replicas.asciidoc [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    "index" : {
+        "number_of_shards" : 1,
+        "number_of_replicas" : 4,
+        "data_path": "/var/data/my_index",
+        "shadow_replicas": true
+    } 
+}'
+--------------------------------------------------
+
+[WARNING]
```
Comment (s1monw) [inline docs/reference/indices/shadow-replicas.asciidoc]: I think we should leave a note about the limitation of the replicas being behind mapping wise?

File: src/main/java/org/elasticsearch/index/IndexService.java [hunk 5]
```diff
@@ -304,15 +304,16 @@ public synchronized IndexShard createShard(int sShardId) throws ElasticsearchExc
             indicesLifecycle.beforeIndexShardCreated(shardId, indexSettings);
 
             logger.debug("creating shard_id {}", shardId);
-
+            // if we are on a shared FS we only own the shard (ie. we can safely delete it) if we are the primary.
+            final boolean ownsShard = IndexMetaData.usesSharedFilesystem(indexSettings) == false || (primary && IndexMetaData.usesSharedFilesystem(indexSettings));
```
Comment (s1monw) [inline src/main/java/org/elasticsearch/index/IndexService.java]: I wonder if we should name this after what we use it for ie. `canDeleteShardContent`?

File: src/main/java/org/elasticsearch/index/engine/ShadowEngineFactory.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.engine;
+
+/**
+ * TODO: document me!
```
Comment (s1monw) [inline src/main/java/org/elasticsearch/index/engine/ShadowEngineFactory.java]: either remove or add a short docstring

Comment (s1monw) [inline src/main/java/org/elasticsearch/index/engine/ShadowEngineFactory.java]: I wonder if we should extend EngineFactory to have two methods `newReadOnlyEngine` / `newReadWriteEngine` or pass a flag to the `newEngine(Type, EngineConfig)` I think it's cleaner than having a dedicated EngineFacotory?

File: src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java [hunk 7]
```diff
@@ -136,6 +138,10 @@ public void startRecovery(final IndexShard indexShard, final RecoveryState.Type
             logger.debug("{} ignore recovery. already in recovering process, {}", indexShard.shardId(), e.getMessage());
             return;
```
Comment (s1monw) [inline src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java]: I think we can revert the changes here entirely? they are all lefovers

File: src/test/java/org/elasticsearch/index/IndexWithShadowReplicasTests.java [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertHitCount;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.greaterThanOrEqualTo;
+
+/**
+ * Tests for indices that use shadow replicas and a shared filesystem
+ */
+@ElasticsearchIntegrationTest.ClusterScope(scope = ElasticsearchIntegrationTest.Scope.TEST, numDataNodes = 0)
+public class IndexWithShadowReplicasTests extends ElasticsearchIntegrationTest {
+
```
Comment (s1monw) [inline src/test/java/org/elasticsearch/index/IndexWithShadowReplicasTests.java]: I think we have to have at least one test that uses FieldData to ensure the reader handling is all correct etc?

**Comment by s1monw** [comment]
@dakrone I left a bunch of comments

**Comment by s1monw** [comment]
@dakrone I think we have to implement a MockShadowEngine as well that does similar things as the MockInternalEngine does. Can you do that? I think it can share quite some code?

File: src/main/java/org/elasticsearch/index/engine/ShadowEngineFactory.java (hunk 6 above)
Comment (dakrone) [inline src/main/java/org/elasticsearch/index/engine/ShadowEngineFactory.java]: I agree, I'll change the `EngineFactory`

File: src/main/java/org/elasticsearch/index/IndexService.java (hunk 5 above)
Comment (dakrone) [inline src/main/java/org/elasticsearch/index/IndexService.java]: Sure, I will rename this.

File: docs/reference/indices/shadow-replicas.asciidoc (hunk 4 above)
Comment (dakrone) [inline docs/reference/indices/shadow-replicas.asciidoc]: I'll add a section to the documentation on this.

File: src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java [hunk 9]
```diff
@@ -631,9 +631,22 @@ void performOnReplica(final ReplicationState state, final ShardRouting shard, fi
             }
 
             final ReplicaOperationRequest shardRequest = new ReplicaOperationRequest(shardIt.shardId(), state.replicaRequest());
+
+            // If the replicas use shadow replicas, there is no reason to
+            // perform the action on the replica, so skip it and
+            // immediately return
+            if (IndexMetaData.isIndexUsingShadowReplicas(indexMetaData.settings())) {
+                // this doesn't replicate mappings changes, so can fail if mappings are not predefined
+                // It was successful on the replica, although we never actually executed - in the future we will
```
Comment (kimchy) [inline src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java]: I would clarify this statement? cause mapping updates do get replicated, it just takes longer since it needs to head to the master and then published to the replicas, so there is a delay in mapping introduction.

File: src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java [hunk 10]
```diff
@@ -784,4 +786,23 @@ public static void writeTo(IndexMetaData indexMetaData, StreamOutput out) throws
             }
         }
     }
+
+    /**
+     * Returns <code>true</code> iff the given settings indicate that the index associated
+     * with these settings allocates it's shards on a shared filesystem. Otherwise <code>false</code>. The default
+     * setting for this is the returned value from {@link #isIndexUsingShadowReplicas(org.elasticsearch.common.settings.Settings)}.
+     */
+    public static boolean usesSharedFilesystem(Settings settings) {
```
Comment (kimchy) [inline src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: can we use the same method structure between this method and the following? I like `isIndex...`

File: src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java (hunk 7 above)
Comment (dakrone) [inline src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java]: Will do.

File: src/test/java/org/elasticsearch/index/IndexWithShadowReplicasTests.java (hunk 8 above)
Comment (dakrone) [inline src/test/java/org/elasticsearch/index/IndexWithShadowReplicasTests.java]: I'll add a test

File: src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java (hunk 9 above)
Comment (dakrone) [inline src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java]: I will clarify this comment

File: src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java (hunk 10 above)
Comment (dakrone) [inline src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: Yep, I'll rename.

**Comment by dakrone** [comment]
@s1monw I added a `MockShadowEngine` and addressed your other comments

File: src/main/java/org/elasticsearch/index/shard/IndexShardModule.java [hunk 11]
```diff
@@ -19,34 +19,60 @@
...
-    public IndexShardModule(ShardId shardId, Settings settings) {
+    public IndexShardModule(ShardId shardId, boolean primary, Settings settings) {
         this.settings = settings;
         this.shardId = shardId;
+        this.primary = primary;
+    }
+
+    /** Return true if a shadow engine should be used */
+    protected boolean useShadowEngine() {
+        return primary == false && settings.getAsBoolean(IndexMetaData.SETTING_SHADOW_REPLICAS, false);
```
Comment (kimchy) [inline src/main/java/org/elasticsearch/index/shard/IndexShardModule.java]: should we use the `IndexMetaData#isIndexUsingShadowReplicas` help method here?

**Comment by kimchy** [comment]
left really minor comments, it looks great. One note, should we mention in the docs the second phase tasks, like do primary promotion without failing an engine? If so, I would also add a task that on get, we automatically set the "go to primary" flag if shadow replica is used and realtime get is used?

File: src/main/java/org/elasticsearch/index/shard/IndexShardModule.java (hunk 11 above)
Comment (dakrone) [inline src/main/java/org/elasticsearch/index/shard/IndexShardModule.java]: Yeah, I will change this to use the helper

**Comment by s1monw** [comment]
left two more comments other than that LGTM

**Comment by dakrone** [comment]
Pushed more commits hooking up the `MockShadowEngine`, moving the engine creation into the `ShadowIndexShard` and automatically using `?preference=_primary` when doing a realtime GET.

File: src/main/java/org/elasticsearch/action/get/TransportGetAction.java [hunk 12]
```diff
@@ -74,6 +75,14 @@ protected void resolveRequest(ClusterState state, InternalRequest request) {
         if (request.request().realtime == null) {
             request.request().realtime = this.realtime;
         }
+        IndexMetaData indexMeta = state.getMetaData().index(request.concreteIndex());
+        if (request.request().realtime && // if the realtime flag is set
+                request.request().preference() == null && // the preference flag is not already set
+                indexMeta != null && // and we have the index
+                IndexMetaData.isIndexUsingShadowReplicas(indexMeta.settings())) { // and the index uses shadow replicas
+            // set the preference for the request to use "_primary" automatically
+            request.request().preference("_primary");
```
Comment (s1monw) [inline src/main/java/org/elasticsearch/action/get/TransportGetAction.java]: can we use `org.elasticsearch.cluster.routing.operation.plain.Preference.PRIMARY.type()` here instead?

**Comment by s1monw** [comment]
left on minor comment! LGTM feel free to push!

File: docs/reference/indices/shadow-replicas.asciidoc [hunk 13]
```diff
@@ -0,0 +1,105 @@
+[[indices-shadow-replicas]]
+== Shadow replica indices
+
+experimental[]
```
Comment (jpountz) [inline docs/reference/indices/shadow-replicas.asciidoc]: Thanks for this!

**Comment by dakrone** [comment]
pushed to 1.x and master!
